go/pkg/tool/tool.go

```diff
@@ -848,3 +858,92 @@ func (c *Client) getActionResult(ctx context.Context, actionDigest string) (*rep
 	}
 	return resPb, nil
 }
+
+// IO is a collection input root Digest, Inputs, and Outputs for an action.
+type IO struct {
+	RootDg  string
+	Inputs  *Inputs
+	Outputs *Outputs
+}
+
+// Inputs are input Paths (with digests) of an action. For a symlink, the key is
+// `<path>-><target>`, the value is the dg of the target file.
+type Inputs struct {
+	Paths        map[string]string
+	PathSymlinks map[string]string
+}
+
+// Outputs are output Files/Dirs (with digests) of an action. For a symlink, the
+// key is `<path>-><target>`, the value is the dg of the target file.
+type Outputs struct {
+	Files        map[string]string
+	Dirs         map[string]string
+	FileSymlinks map[string]string
+	DirSymlinks  map[string]string
+}
+
+// GetIO returns the Inputs and Outputs of an action Digest.
+func (c *Client) GetIO(ctx context.Context, actionDigest string) (*IO, error) {
+	acDg, err := digest.NewFromString(actionDigest)
+	if err != nil {
+		return nil, fmt.Errorf("error creating action digest: %w", err)
+	}
+	actionProto := &repb.Action{}
+	// Get all the Inputs.
+	if _, err := c.GrpcClient.ReadProto(ctx, acDg, actionProto); err != nil {
+		return nil, fmt.Errorf("error reading from proto: %w", err)
+	}
+	rootDg, err := digest.NewFromProto(actionProto.GetInputRootDigest())
+	if err != nil {
+		return nil, fmt.Errorf("error getting input root digest: %w", err)
+	}
+	_, inputPaths, inputSymlinks, err := c.getInputTree(ctx, actionProto.GetInputRootDigest())
+	if err != nil {
+		return nil, err
+	}
+	// Get all the Outputs.
+	resPb, err := c.getActionResult(ctx, actionDigest)
+	if err != nil {
+		return nil, err
```

Comment:
For actions that failed, they wouldn't have an action result. Would it be better to just add a warning in case getting action result fails and leave the outputs as empty?